File: src/patterns/packagejson.ts
import type { Extractor, ExtractorFn } from "./extractor.js"
import type { GlobRule } from "./rule.js"

import { npmManifestParse } from "../targets/npmManifest.js"
import { ruleCompile } from "./resolveSources.js"
import { resolveNegatable, type Source } from "./source.js"

const decoder = new TextDecoder()

/**
 * Extracts and compiles patterns from the file.
 *
 * @see {@link ruleCompile}
 *
 * @since 0.6.0
 */
export function extractPackageJson(source: Source, content: Uint8Array): void | null | Error {
	try {
		return extractPackageJsonRules(source, content)
	} catch (e) {
		return e as Error
	}
}

extractPackageJson satisfies ExtractorFn

export function makePackageJsonExtractor(
	mode: "list" | "publish" | "bundle" = "publish",
): Extractor {
	return {
		extract(source, content) {
			try {
				return extractPackageJsonRules(source, content, mode)
			} catch (e) {
				return e as Error
			}
		},
		path: "./package.json",
	}
}

export const packageJsonExtractor: Extractor = makePackageJsonExtractor("publish")

/**
 * Extracts and compiles patterns from the file.
 *
 * @since 0.12.0
 */
export function extractPackageJsonRules(
	source: Source,
	content: Uint8Array,
	mode: "list" | "publish" | "bundle" = "publish",
): void | null {
	let dist: { files?: string[] }

	try {
		dist = npmManifestParse(decoder.decode(content), mode)
	} catch (err) {
		throw new Error("Invalid '" + source.path + "'", { cause: err })
	}

	if (!dist || !Array.isArray(dist.files)) return null

	source.inverted = true
	let rule: GlobRule | undefined

	const options = { nocase: true }
	for (const pattern of dist.files) {
		const nextRule = resolveNegatable(pattern, true, rule)
		if (nextRule === rule) continue
		rule = nextRule
		source.rules.unshift(rule)
	}

	const rlen = source.rules.length
	for (let i = 0; i < rlen; i++) {
		const r = source.rules[i]!
		if ("list" in r && r.compiled === null) ruleCompile(r, options)
	}
}

File: internal/patterns/packagejson.go
package patterns

import (
	"encoding/json"
)

const ExtractorError ExtractorNext = 2

type NodeJsManifest struct {
	Files *[]string `json:"files"`
}

func ExtractPackageJson(source *Source, content []byte, _ *MatcherContext) ExtractorNext {
	result := extractPackageJson(source, content)
	if result == ExtractorBreak {
		for element := range source.Pattern {
			element.Compile(StringCompileOptions{NoCase: true})
		}
	}
	if result == ExtractorError {
		return ExtractorBreak
	}
	return result
}

var _ ExtractorFn = (ExtractorFn)(ExtractPackageJson)

func extractPackageJson(source *Source, content []byte) ExtractorNext {
	source.Inverted = true
	include := SignedPattern{}
	exclude := SignedPattern{Excludes: true}
	dist := NodeJsManifest{}
	err := json.Unmarshal(content, &dist)
	if err != nil {
		source.Error = err
		return ExtractorError
	}

	if dist.Files == nil {
		return ExtractorBreak
	}

	for _, pattern := range *dist.Files {
		source.PushNegatable(pattern, true, include, exclude)
	}
	source.Pattern = append(source.Pattern, include, exclude)

	return ExtractorBreak
}
